### PR TITLE
Add bucket type/index n_val equivalency validator

### PR DESCRIPTION
**Pull request by coderoshi**
Registers a create/update bucket type property validator for yokozuna. This ensures that setting properties on a bucket type will validate that 1) the index exists, and 2) the bucket types n_val equals the index's n_val.

This forced a change to the tests, whereby an index must first be created, then `wait_for_index`, then associate the index with the bucket. Before this check, waiting didn't matter until values were written.

### CODE REVIEW

**Comment by rzezeski** [comment]
- [x] code review
- [x] compile/eunit
- [x] dialyzer
- [x] riak tests
- [x] manual test

File: riak_test/yokozuna_essential.erl [hunk 1]
```diff
@@ -269,19 +269,18 @@ setup_indexing(Cluster, PBConns, YZBenchDir) ->
...
     ok = yz_rt:create_index(Node, <<"escaped">>),
-    yz_rt:set_index(Node, {?BUCKET_TYPE, <<"escaped">>}, <<"escaped">>),
-
     ok = yz_rt:create_index(Node, <<"unique">>),
+
     [yz_rt:wait_for_index(Cluster, I)
-     || I <- [?INDEX, <<"tagging">>, <<"escaped">>, <<"unique">>]].
+     || I <- [?INDEX, <<"tagging">>, <<"escaped">>, <<"unique">>]],
+
+    ok = rpc:call(Node, riak_core_bucket, set_bucket,
```
Comment (rzezeski) [inline riak_test/yokozuna_essential.erl]: This `rpc:call` shouldn't be needed anymore now that you are passing `NVal` to `set_index`.

File: src/yz_bucket_validator.erl [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            %% since we cannot remove n_val
+            {proplists:delete(search_index, BucketProps),
+                [{search_index, Error}]}
+    end.
+
+-spec get_search_index_info(props()) -> {error,binary()} | {index_name(),n()}.
+get_search_index_info(Props) ->
+    case proplists:get_value(search_index, Props) of
+        undefined ->
+            {error, no_search_index};
```
Comment (rzezeski) [inline src/yz_bucket_validator.erl:79]: This doesn't match your type spec:

```
yz_bucket_validator.erl:51: The pattern {'error', 'no_search_index'} can never match the type {'error' | binary(),binary() | pos_integer()}
```

File: src/yz_bucket_validator.erl [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                false ->
+                    {error, ?FMT("~s does not exist", [Index])};
+                true ->
+                    {Index, index_n_val(Index)}
+            end
+    end.
+
+%% @doc Get search_index from the bucket props, and return the
+%%      index's n_val. If it doesn't exist, return undefined.
+% index_n_val(Props) ->
```
Comment (rzezeski) [inline src/yz_bucket_validator.erl]: Commented out code, also add `@priavte`.

File: src/yz_bucket_validator.erl [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            Error = ?FMT("Bucket type n_val ~p must match the associated "
+                         "search_index ~s n_val ~p ", [BNVal,Index,INVal]),
+            %% Removing search_index from valid props,
+            %% since we cannot remove n_val
+            {proplists:delete(search_index, BucketProps),
+                [{search_index, Error}]}
+    end.
+
+-spec get_search_index_info(props()) -> {error,binary()} | {index_name(),n()}.
+get_search_index_info(Props) ->
```
Comment (rzezeski) [inline src/yz_bucket_validator.erl:76]: Add `@private`.

File: src/yz_bucket_validator.erl [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            {proplists:delete(search_index, BucketProps),
+                [{search_index, Msg}]};
+        {Index, INVal} ->
+            BNVal = proplists:get_value(n_val, Props),
+            validate_n_val(Index, INVal, BNVal, BucketProps)
+    end.
+
+%% @private
+%%
+
```
Comment (rzezeski) [inline src/yz_bucket_validator.erl]: Dat whitespace.

File: src/yz_bucket_validator.erl [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+-define(FMT(S, Args), list_to_binary(lists:flatten(io_lib:format(S, Args)))).
+
+-type prop() :: {PropName::atom(), PropValue::any()}.
+-type error() :: {PropName::atom(), ErrorReason::atom()}.
+-type props() :: [prop()].
+-type errors() :: [error()].
+
+%% @doc Performs two validations. The first is validating that an index
+%% exists before a bucket type can be associated to it by setting search_index.
```
Comment (rzezeski) [inline src/yz_bucket_validator.erl]: This applies to buckets in general, not just bucket types, right?

**Comment by rzezeski** [comment]
This doesn't seem to actually be working. Here are steps from manual
test I did:

```
# curl -X PUT 'http://localhost:10018/search/index/foo'

# curl 'http://localhost:10018/search/index/foo' | json_pp
{
   "n_val" : 3,
   "name" : "foo",
   "schema" : "_yz_default"
}

# curl -i -X PUT -H 'content-type: application/json' 'http://localhost:10018/types/data/buckets/foo/props' -d '{"props":{"search_index":"foo"}}'

# curl 'http://localhost:10018/types/data/buckets/foo/props' | json_pp
{
   "props" : {
      ...
      "n_val" : 3,
      ...
      "search_index" : "foo",
   }
}

# curl -i -X PUT -H 'content-type: application/json' 'http://localhost:10018/types/data/buckets/foo/props' -d '{"props":{"n_val":4}}'
HTTP/1.1 204 No Content
Vary: Accept-Encoding

// What? That shouldn't have work.

# curl 'http://localhost:10018/types/data/buckets/foo/props' | json_pp
{
   "props" : {
      ...
      "search_index" : "foo",
      ...
      "n_val" : 4,
      ...
   }
}

// This should not have been possible. There is nothing in the logs to
// indicate error. Then for heck of it I try again with n_val:5

# curl -i -X PUT -H 'content-type: application/json' 'http://localhost:10018/types/data/buckets/foo/props' -d '{"props":{"n_val":5}}'
HTTP/1.1 400 Bad Request

// Okay that one failed, I check the logs:

==> /root/yz-verify/rt/riak_yz/dev/dev1/log/console.log <==
2014-02-14 17:02:24.472 [error] <0.1688.0>@riak_core_bucket:set_bucket:77 Bucket properties validation failed [{search_index,<<"Bucket type n_val 4 must match the associated search_index foo n_val 3 ">>}]

// Interesting so even though I set to 5 it's reading the current
// value of 4. Something strange is happening with this validator code.
```

We should add two tests to `yz_index_admin`.
1. That non-existing index cannot be associated to bucket.
2. That n_val must match of the set property should fail.

File: src/yz_bucket_validator.erl [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        {ExistingProps, undefined} ->
+            ExistingProps;
+        {ExistingProps, BucketProps} ->
+            riak_core_bucket_props:merge(ExistingProps, BucketProps)
+    end,
+    case get_search_index_info(Props) of
+        {error, no_search_index} ->
+            {BucketProps, []};
+        {error, Msg} ->
+            {proplists:delete(search_index, BucketProps),
```
Comment (rzezeski) [inline src/yz_bucket_validator.erl]: I haven't read up on how validators work so I may be misunderstanding what this code does but I hope this is not deleting the `search_index` property from the bucket properties. If the validation fails it should fail the user request and not mutate the stored properties.

Comment (coderoshi) [inline src/yz_bucket_validator.erl]: A tuple of {valid properties, error messages} are to be returned. We aren't mutating the properties here, we're only removing search_index from the list of acceptable ones. See `riak_kv_bucket:validate/4`.

**Comment by rzezeski** [comment]
Riak Tests pass and manual test works:

```
curl -i -X PUT -H 'content-type: application/json' 'http://localhost:10018/types/data/buckets/foo/props' -d '{"props":{"n_val":4}}'
HTTP/1.1 400 Bad Request
Vary: Accept-Encoding
Server: MochiWeb/1.1 WebMachine/1.10.5 (jokes are better explained)
Date: Sat, 15 Feb 2014 01:14:53 GMT
Content-Type: application/json
Content-Length: 82

{"n_val":"Bucket type n_val 4 must match the associated search_index foo n_val 3"}
```

I think the error message should be updated to just say "Bucket" but other than that looks good.

+1 to merge.

**Pull request merged**